SI-6240 synchronization for runtime reflection

xeno-by opened the pull request:
Locks are no fun, but it looks like that's our only hope of getting thread-safety in 2.10.1.

xeno-by committed:
synchronizes runtimeMirror

xeno-by committed:
synchronizes PackageScope.negatives

We don't have to serialize access to this cache, but we ought to
make the cache safe for being accessed from multiple threads at once.

xeno-by committed:
synchronizes toolboxes

We don't guarantee thread-safety of the front end, but everything else
looks good now.

xeno-by committed:
synchronizes symbol materializers

Since runtime reflection doesn't have a luxury of enumerating all classes
on the classpath, it has to materialize symbols for top-level definitions
(packages, classes, objects) on demand.

Someone asks us for a class named `foo.Bar`? Easy. Let's speculatively create
a package named `foo` and then look up `newTypeName("bar")` in its decls.
This lookup, implemented in `SymbolLoaders.PackageScope` tests the waters by
trying to to `Class.forName("foo.Bar")` and then creates a ClassSymbol upon
success (the whole story is a bit longer, but the rest is irrelevant here).

That's all neat, but these non-deterministic mutations of the global symbol
table give a lot of trouble in multi-threaded setting. One of the popular
reflection crashes happens when multiple threads happen to trigger symbol
materialization multiple times for the same symbol, making subsequent
reflective operations stumble upon outrageous stuff like overloaded packages.

Short of significantly changing SymbolLoaders I see no other way than just
to slap a global lock on materialization in runtime reflection.

xeno-by committed:
runtime reflection: death from thousand threads

not anymore

xeno-by committed:
removes the assertion in missingHook

In the current synchronization scheme multiple threads can enter the
missingHook trying to materialize a package, which hasn't been created.

That's fine, because makeScalaPackage, which creates and enters package
symbols is synchronized and checks whether the creation is necessary
before commencing. Therefore even if makeScalaPackage is called multiple
times in rapid succession, the calls will be serialized and all calls
except the first one won't do anything.

xeno-by committed:
cleans up initialization of runtime reflection

At first I just tried to remove syntheticCoreClasses from missingHook
and put them into the initializer of freshly created mirrors in order to
reduce the non-determinism in mutations of the global symbol table.

And then it didn't work, crashing on me claiming that AnyRef is missing.
Apparently we still need AnyRefClass in missingHook, just because it's
impossible to initialize (i.e. unpickle) ScalaPackageClass without it.

And then it still didn't work, whining about multiple overloaded defs
of some synthetic symbols. That was really tricky, but I figured it out
as well by initializing ScalaPackageClass first before forcing any
synthetic symbols (see the details in comments).

xeno-by committed:
introducing GIL to Scala reflection

On a serious note, I feel really uncomfortable about having to juggle
this slew of locks. Despite that I can't immediately find a deadlock,
I'm 100% sure there is one hiding in the shadows. Hence, I'm abandoning
all runtime reflection locks in favor of a single per-universe one.

xeno-by committed:
reflection no longer uses atPhase and friends

Mentioned methods mutate the global `atPhaseStack` variable, which can
easily lead to imbalances and, ultimately, to the empty stack error.

Luckily for us, there's only one dummy phase, SomePhase, which is used
by runtime reflection, so there is absolutely zero need to invoke atPhase
in non-compiler reflexive universes.

xeno-by committed:
synchronizes symbols

Synchronization via decoration would be neat if it actually worked.
Unfortunately, root symbols never got decorated, therefore their children
also never got decorated and all the way down to the very turtles.

This commit fixes this sad issue by turning root symbols from objects
to lazy vals. Yes, this is going to induce a performance penalty, which
will hopefully not be high enough to invalidate this cornerstone of our
synchronization strategy.

Now when root symbols are lazy vals, they can be overridden in the runtime
reflexive universe and decorated with SynchronizedSymbol, which makes their
children sync and sound.

xeno-by commented:
review @retronym @paulp @adriaanm @gkossakowski

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1727/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2462/

xeno-by committed:
SI-6539 moves @compileTimeOnly away from scala-reflect

The move is done to provide forward compatibility with 2.10.0.

The annotation isn't replaced with one of the macro-based solutions
right away (see comments for more information about those), because
we lack necessary tech in 2.10.x.

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1727/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2465/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2462/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2465/

xeno-by commented:
We discussed the locking matter with Roland and Martin, and it looks like it might be possible to get rid of `Symbol.info` and completer locks altogether. There will still be a necessity for locks in `MethodSymbol.typeAsMemberOf` and `TypeSymbol.tpe`, but the uses of those are much more rare. (We could also get rid of them later by making cache entries atomic if performance permits).

Other locks we have include `nameLock` for generating names (probably can be replaced with a CAS), `baseTypeSeqLock` for BaseTypeSeqs, which involve refined types (rare!) and all sort of locks in `SynchronizedTypes.scala`.

The latter is a huge problem, because such widely used method as `<:<` must be protected with a GIL, because it involves mutable global vars in the compiler. I plan on looking into that tomorrow.

xeno-by commented:
Having said all that, I suggest we merge this pull request as it is, which would allow me to build upon it to elide locks in `info` and `<:<` (two blockers outlined by Roland).

I see two blockers for that:
1) Getting reviews.
2) Proving that changing root symbols from `object` to `lazy val` in ffc0c5f doesn't degrade performance of the compiler. What would you advise here?

/cc @adriaanm @gkossakowski @retronym @paulp @rkuhn

paulp commented:
Have you made friends with the undolog? I wish you great luck in pitching that one into a ravine.

xeno-by commented:
Luckily someone made undolog synchronizable before me. Optimizing the performance of synchronization would be a completely different thing, of course.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1795/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2531/

xeno-by commented:
Kitty has started building stuff, because I occasionally pushed an extraneous commit. Good job kitty, previously you weren't so responsive! Permanent catnips ftw.

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2531/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1795/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

retronym commented on src/reflect/scala/reflect/runtime/SynchronizedTypes.scala:
```diff
@@ -12,8 +12,8 @@ private[reflect] trait SynchronizedTypes extends internal.Types { self: SymbolTa
 
-  private object uniqueLock
-
+  // we can keep this lock fine-grained, because super.unique do anything extraordinary, which makes deadlocks impossible
```
"because super.unique do anything extraordinary" ?

xeno-by replied:
Fixed.

retronym commented on test/files/run/reflection-sync.scala:
```diff
@@ -0,0 +1,12 @@
+    }
+  })
+  threads foreach (_.start)
+}
```
Can you add comments to show exactly what this test case is designed to stress? Initialization? Normal operation? Can we rig tests to exercise the initialization more heavily? Could this test use a broader set of type tags, perhaps some with inter-dependencies?

I do know these tests are tricky to write.

xeno-by replied:
Done a bit more in the test.

retronym commented on src/reflect/scala/reflect/runtime/Gil.scala:
```diff
@@ -0,0 +1,21 @@
+        gil.unlock()
+      }
+    }
+  }
```
Might it be better to make this `@inline` to allow closure elimination? It would need to be in a class to allow this (e.g. `internal.SymbolTable`). In any case, the method could/should be final.

gkossakowski replied:
@retronym: until we get existing `@inline` to work I don't think it makes sense to add any more.

retronym replied:
Which bug are you referring to? I know trait inlining is not supported, but we rely on non-trait inlining in many other places of the compiler already for methods that look a lot like this one.

gkossakowski replied:
@retronym: with current master if you build quick using `ant -Dscalac.args="-Yinline-warnings" build-opt` you get:
814 warnings `At the end of the day, could not inline @inline-marked method ...`
258 warnings `Could not inline required method ...`

That's 1072 inliner warnings that nobody understands. If we keep adding `@inline` without looking addressing existing warnings we just make the situation worse, not better.

Here's the log: https://gist.github.com/4687953

retronym replied:
Right, but even though there are 4 warnings about `degubLog`, about 40 other call sites do inline.

There are a bunch of calls that documented not to work:

```
   * Forcing direct fields access using the @inline annotation helps speed up
   * various operations (especially smallest/greatest and update/delete).
   *
   * Unfortunately the direct field access is not guaranteed to work (but
   * works on the current implementation of the Scala compiler).
   *
   * An alternative is to implement the these classes using plain old Java code...
   */
  sealed abstract class Tree[A, +B](
    @(inline @getter) final val key: A,
    @(inline @getter) final val value: B,
    @(inline @getter) final val left: Tree[A, B],
    @(inline @getter) final val right: Tree[A, B])
```

`any2arrowassoc` also contributes an enourmous number, but that's always been troublesome.

I've got to run, so I can't comment more right now. But let's find a way to avoid these closures; if we want guaranteed inlining we could also do it with a macro.

xeno-by replied:
I'm marking the method as `@inline final` and leaving it in `Gil.scala`. In this case, I think, clarity of the file where the code resides trumps performance, which isn't even a problem for now.

retronym commented on src/reflect/scala/reflect/runtime/Gil.scala:
```diff
@@ -0,0 +1,21 @@
+  self: SymbolTable =>
+
+  // fixme... please...
+  lazy val gil = new java.util.concurrent.locks.ReentrantLock
```
Please elaborate in the comment (ticket/email-thread reference). I know you intend to revisit this personally, but... you know.

xeno-by replied:
Done

retronym commented on src/reflect/scala/reflect/internal/StdNames.scala:
```diff
@@ -265,6 +265,8 @@ trait StdNames {
     final val SourceFileATTR: NameType             = "SourceFile"
     final val SyntheticATTR: NameType              = "Synthetic"
 
+    final val scala: NameType = "scala"
```
Identifiers called `scala` give me the shivers; the compiler still fails on hygiene in a few places.

xeno-by replied:
How should this be called?

retronym replied:
`scala_` or `scalaName`.

Just want to guard against confusion in places like:

``` scala
// Mixin
def bitmapCategory(field: Symbol): Name = {
     import nme._
     ...
```

xeno-by replied:
Done

retronym commented on src/reflect/scala/reflect/internal/Definitions.scala:
Can we strengthen this with an assertion in some strategic location that would detect if we somehow ended up with an unpickled `ScalaPackageClass`?

xeno-by committed:
optimizes Scala reflection GIL

First of all, GIL should only apply to runtime reflection, because noone
is going to run toolboxes in multiple threads: a) that's impossible, b/c
the compiler isn't thread safe, b) ToolBox api prevents that.

Secondly, the only things in symbols which require synchronization are:
1) info/validTo (completers aren't thread-safe),
2) rawInfo and its dependencies (it shares a mutable field with info)
3) non-trivial caches like in typeAsMemberOfLock

If you think about it, other things like sourceModule or associatedFile
don't need synchronization, because they are either set up when a symbol
is created or cloned or when it's completed.

We can say that symbols can be in four possible states: 1) being created,
2) created, but not yet initialized, 3) initializing, 4) initialized.
single thread. #2 and #4 don't need synchronization either, because the
only mutation symbols in runtime reflection can undergo is init. #3 is
dangerous and needs protection.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1835/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2570/

xeno-by commented on src/reflect/scala/reflect/internal/Definitions.scala:
Introduced an assert in `PackageScope.enter`, which prevents entering the same symbol twice, since what we ultimately want to achieve here is to avoid duplicates.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1836/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2571/

xeno-by commented:
Accommodated the feedback, rebased off fresh 2.10.x. PLS REBUILD ALL

xeno-by commented:
Oh nice, now I'm tripping my own assertion in my own test. Looks like I have to put off the fix till tomorrow morning.

scala-jenkins commented:
jenkins job pr-rangepos: SUCCESS - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1836/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1836/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2572/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2572/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: ABORTED - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2572/
![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)
SPURIOUS ABORT? -- PLS REBUILD pr-scala-testsuite-linux-opt

xeno-by committed:
Merge pull request #2019 from scalamacros/ticket/6539

SI-6539 moves @compileTimeOnly away from scala-reflect

xeno-by commented:
Needs some work, will reopen soon.

xeno-by commented:
Weird. Github didn't pick up the update to 2.10.x.

scala-jenkins commented:
jenkins job pr-rangepos: SUCCESS - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1836/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: ABORTED - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2572/
![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)
tried automatically rebuilding once before, not falling for it again!
